### PR TITLE
Added Ollama as an inference impl 

### PR DESCRIPTION
**Pull request by hardikjshah**
## Inline Inference 
- Fixed non-streaming inference sever/client/event_logger code 
- Added tests for inline inference to verify that things are working 
`python -m unittest tests/test_inference.py`

## Ollama Inference
- Added non-streaming inference using ollama python apis 
- Added tests to verify that ollama inference can implement the inference api spec across tools, custom tools, etc.
- Streaming inference and appropriate tests 
- [TBD] Setup instructions for ollama and configuration 
- [TBD] pass right request options like temperature, etc

### CODE REVIEW

File: llama_toolchain/inference/api_instance.py [hunk 1]
```diff
@@ -12,6 +12,10 @@ async def get_inference_api_instance(config: InferenceConfig):
         from .inference import InferenceImpl
 
         return InferenceImpl(config.impl_config)
+    elif config.impl_config.impl_type == ImplType.ollama.value:
+        from .inference import OllamaInference
```
Comment (ashwinb) [inline llama_toolchain/inference/api_instance.py:16]: looks to be a typo. this should be `from .ollama`

File: llama_toolchain/inference/ollama.py [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+                    yield ChatCompletionResponseStreamChunk(
+                        event=ChatCompletionResponseEvent(
+                            event_type=ChatCompletionResponseEventType.progress,
+                            delta=ToolCallDelta(
+                                content="",
+                                parse_status=ToolCallParseStatus.started,
+                            ),
+                        )
+                    )
+                    buffer = buffer[len("<|python_tag|>") :]
```
Comment (ashwinb) [inline llama_toolchain/inference/ollama.py:127]: in this case `buffer` isn't even being added to?

File: llama_toolchain/inference/ollama.py [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+                event=ChatCompletionResponseEvent(
+                    event_type=ChatCompletionResponseEventType.complete,
+                    delta="",
+                    stop_reason=stop_reason,
+                )
+            )
+
+
+#TODO: Consolidate this with impl in llama-models
+def decode_assistant_message_from_content(
```
Comment (ashwinb) [inline llama_toolchain/inference/ollama.py:201]: do you recall what the differences were and why you needed to copy this?

File: tests/test_inference.py [hunk 4]
```diff
@@ -0,0 +1,307 @@
+# Run this test using the following command:
```
Comment (ashwinb) [inline tests/test_inference.py:1]: these are amazing! thank you for putting down some tests

**Review by ashwinb — APPROVED** [review]

File: llama_toolchain/inference/ollama.py (hunk 3 above)
Comment (hardikjshah) [reply llama_toolchain/inference/ollama.py:201]: one of the main reason being that there is no separate tokenizer , which the method is models takes. 
I ll work on consolidating or documenting the differences , as a fast follow.

File: llama_toolchain/inference/ollama.py (hunk 2 above)
Comment (hardikjshah) [reply llama_toolchain/inference/ollama.py:127]: The assumption is that we the model always starts with `<|python_tag|>` and hence its just taking everything after that tag.

Comment (ashwinb) [reply llama_toolchain/inference/ollama.py:127]: no but `buffer` the variable isn't even being assigned to -- nothing is being to added to it?

File: llama_toolchain/inference/api_instance.py (hunk 1 above)
Comment (hardikjshah) [reply llama_toolchain/inference/api_instance.py:16]: nice catch, updated the test to use this api.

**Pull request merged**